Add views/functions to check missing and orphaned data files

huansong opened the pull request:
Description
--------------

1. Add views to get _existing_ relation files in the database, including the default, global and user tablespaces. Note that these views won't expose files outside of the data directory. We only look into those under `base/`, `global/` and user-defined tablespace directories.
2. Add views to get _expected_ relation files in the database, using the knowledge from the catalog.
3. Using 1 and 2, construct views to get the **_missing_** files (i.e. files that are expected but not existed) and **_orphaned_** files (i.e. files that are there unexpectedly).
4. Create views to run the above views in MPP. Also, we support checking extended data files for AO/CO tables. 
5. Add regress tests.

To use:
```sql
-- checking non-extended files (no '.' sign)
select * from gp_toolkit.gp_check_missing_files;
select * from gp_toolkit.gp_check_orphaned_files;

-- checking all data files including the extended data files (e.g. 12345.1, 99999.2). 
-- These do not count supporting files such as .fsm .vm etc. And currently 
-- we only support checking extended data files for AO/CO tables, not heap.
select * from gp_toolkit.gp_check_missing_files_ext;
select * from gp_toolkit.gp_check_orphaned_files_ext;
```

Note:
* Currently, support checking all the non-extended datafile (the
      ones w/o the extension ".x") and the extended datafiles of AO/CO
      tables. The main reason to separate these two is performance:
      constructing expected file list for AO/CO segments is a slow task,
      so only do that if really required.
* For heap tables, currently have no way to get the expected number
      of datafiles for a certain table: we cannot use pg_relation_size
      because that is in turn dependent on the number of datafiels itself.
      So always skip its extended files for now.
* We may want to add the missing file check to gpcheckcat later. For
     orphaned files I'm not so sure as there could be some false positives
     that I haven't realized.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

ashwinstar commented:
I would prefer if provide the MPP version for these instead of `gp_dist_random` exposed usage. Also, would be nice to callout how it guarantees safety of the reported results specifically for gp_check_orphaned_files.

huansong commented:
> I would prefer if provide the MPP version for these instead of gp_dist_random exposed usage.

Sure will do that.

> Also, would be nice to callout how it guarantees safety of the reported results specifically for gp_check_orphaned_files.

By safety do you mean e.g. not exposing unrelated user files outside of the data directory? This should be guaranteed by only using database facility such as `pg_ls_dir` to get the existing file list, which won't have reach to any files outside the data directories (including the user tablespace directory). Will call that out in the messages. Or if there's other aspect please LMK.

huansong commented:
> > Right now we refer to pg_attribute_encoding.attnum to construct
> > the expected extended file list for AOCO tables. Later we might
> > want to switch to use the pg_attribute_encoding.filenum. When
> > #15074 is merged.
> 
> Do we want to implement this change to use filenum now that the PR has been merged?

Yes we do. Done (just `%s/attnum/filenum`).

Also, I changed the `gp_` functions to views which are simpler. And I moved them into `gp_toolkit` which should be a better fit.

bmdoil commented on src/test/regress/input/gp_check_files.source:
```diff
@@ -0,0 +1,76 @@
+select oid from pg_database where datname = current_database() \gset
+\cd :oid
+
+-- delete the existing file which only includes the heap's datafile
+\! rm -rf *
```
This seems pretty dangerous. Can we be explicit in what we're deleting, or use a regex?

I ran the regress test locally from src/test/regress and it removed everything in my cwd.

```
#src/test/regress on git:missingfiles
$ ll
total 4.0K
-rw-rw-r-- 1 bdoil bdoil   0 Mar 14 20:04 987654
-rw-rw-r-- 1 bdoil bdoil   0 Mar 14 20:04 987654.3
-rw-rw-r-- 1 bdoil bdoil 189 Mar 14 20:04 regression.diffs
```

huansong replied:
Oh I'm sorry about that! I hope you didn't delete anything non-recoverable! 

This is a problem for sure. The reason is that in your regress directory the test tablespace directory has not been created. It is normally created by makefile, along with running installcheck. I had run that not far ago so I had it locally w/o even knowing it. If the directory is not there then we would fail to `cd` into it and instead delete everything from the regress directory.

And the reason that I had to do multiple `cd`'s and `rm *` instead of specifying exact path to remove is that psql doesn't allow us pass variable into the shell escape `\!` (see https://linux.die.net/man/1/psql#:~:text=The%20arguments%20are%20not%20further%20interpreted%2C%20the%20shell%20will%20see%20them%20as%20is), while it allows that in `\cd`.

Now, making this safer by (1) checking if the current directory is expected; (2) delete exact matching of files (number-named).

bmdoil commented on src/test/regress/input/gp_check_files.source:
```diff
@@ -0,0 +1,76 @@
+insert into checkmissing_ao select i,i,i from generate_series(1,100)i;
+insert into checkmissing_co select i,i,i from generate_series(1,100)i;
+
+\! find . -type f -name "*.*" -exec rm {} \;
```
Same here, I don't want anyone to delete their dotfiles if they happen to run pg_regress from their home directory.

huansong commented:
@bmdoil Thanks for reviewing. Force-pushed to resolve rebase conflict.
